File: ScriptCore/Source/PhezuEngine/Core/Vector2.cs
using System;
using System.Runtime.InteropServices;

namespace PhezuEngine {

    [StructLayout(LayoutKind.Sequential)]
    public struct Vector2 {
        public float X;
        public float Y;

        public Vector2(float x, float y) {
            X = x;
            Y = y;
        }
        
        public static Vector2 Zero => new Vector2(0, 0);
        public static Vector2 One => new Vector2(1, 1);
        public static Vector2 Right => new Vector2(1, 0);
        public static Vector2 Up => new Vector2(0, 1);
        public static Vector2 Down => new Vector2(0, -1);
        public static Vector2 Left => new Vector2(-1, 0);
        
        public static float Dot(Vector2 lhs, Vector2 rhs) => lhs.X * rhs.X + lhs.Y * rhs.Y;

        public float Magnitude()
        {
            return MathF.Sqrt(X * X + Y * Y);
        }

        public void Normalize()
        {
            float magnitude = Magnitude();

            if (magnitude < float.Epsilon)
                return;
            
            X /= magnitude;
            Y /= magnitude;
        }

        public static Vector2 operator+(Vector2 a, Vector2 b) {
            return new Vector2(a.X + b.X, a.Y + b.Y);
        }

        public static Vector2 operator-(Vector2 a, Vector2 b) {
            return new Vector2(a.X - b.X, a.Y - b.Y);
        }

        public static Vector2 operator*(Vector2 a, float b)
        {
            return new Vector2(a.X * b, a.Y * b);
        }
    }
}

File: Phezu/Source/Maths/Objects/Vector2.cpp
#include "Maths/Objects/Vector2.hpp"
#include "Maths/Objects/Vector3.hpp"

namespace Phezu {

    const Vector2 Vector2::One = Vector2(1.0f, 1.0f);
    const Vector2 Vector2::Zero = Vector2(0.0f, 0.0f);
    const Vector2 Vector2::Right = Vector2(1.0f, 0.0f);
    const Vector2 Vector2::Left = Vector2(-1.0f, 0.0f);
    const Vector2 Vector2::Up = Vector2(0.0f, 1.0f);
    const Vector2 Vector2::Down = Vector2(0.0f, -1.0f);

    Vector2::Vector2(const Vector3& vec) : m_Vector(vec.X(), vec.Y()) {}

    void Vector2::Normalize() {
        if (glm::length(m_Vector) == 0)
            return;
        m_Vector = glm::normalize(m_Vector);
    }

    float Vector2::SqrDistance(const Vector2& a, const Vector2& b) {
        float dx = a.X() - b.X();
        float dy = a.Y() - b.Y();

        return (dx * dx) + (dy * dy);
    }
}